Add missing security annotations

rokaszygmantas opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.6.x
| Description?  | Adding missing security annotations that will check if employee has sufficient permissions before allowing to access the actions.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | Check if specified actions are working with given permissions and showing an error message with insufficient permissions. The actions are listed below.

Actions that are secured with this PR:
* admin_suppliers_search - **search action in suppliers grid**
* admin_suppliers_export - **export action in suppliers grid**
* admin_customers_filter - **search action in customers grid**
* admin_customers_create - **customer creation form page and submit action**
* admin_customers_edit - **customer edit form page and submit action**
* admin_customers_export - **export action in customers grid**
* admin_stock_overview - **stock page `/admin-dev/index.php/sell/stocks/`**
* admin_stock_movements_overview - **stock movements page `/admin-dev/index.php/sell/stocks/movements`**
* admin_import_file_upload - **import file upload action**
* admin_import_sample_download - **import sample file download action**
* admin_employees_search - **search action in employees grid**
* admin_profiles_search - **search action in profiles grid**
* admin_international_translation_overview - **translations app (accessed by clicking "modify" in International>Translations) e.g. `/admin-dev/index.php/improve/international/translations/?lang=en&type=back&locale=en-US`**
* admin_languages_search - **search action in languages grid**
* admin_currencies_search - **search action in currencies grid**
* admin_cms_pages_search_cms_category - **search action in CMS category grid**
* admin_import_data_configuration_get - **functionality from import step 2, which is hidden and still WIP**
* admin_import_get_available_fields - **action to retrieve import available fields, which are displayed in import step 1 (https://prnt.sc/nc4nr4)**

---

Actions for which security is **not** fixed in this PR:
* many actions from product page - @matks informed me that product page will be reworked soon.
* admin_common_recommended_modules - couldn't find where this action is used at all. https://github.com/PrestaShop/PrestaShop/blob/5eaef1459a8d3ceb35ffa0a312a1632877947e9e/src/PrestaShopBundle/Controller/Admin/CommonController.php#L157
* Category page actions - they are fixed in https://github.com/PrestaShop/PrestaShop/pull/13236
* Actions that have specific security rules which are defined in action's code instead of annotations.
* JSON actions - we need to improve security annotations to support JSON responses. It is done in https://github.com/PrestaShop/PrestaShop/pull/13236

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/13372)
<!-- Reviewable:end -->

rokaszygmantas committed:
Add missing security annotations

rokaszygmantas committed:
Add security annotations to ajax actions

rokaszygmantas committed:
Remove unneeded demo annotation

matks commented on src/PrestaShopBundle/Controller/Admin/CommonController.php:
```diff
@@ -330,6 +331,8 @@ public function renderFieldAction($formName, $formType, $fieldName, $fieldData)
      * @param string $redirectRoute
      * @param array $redirectQueryParamsToKeep
      *
+     * @AdminSecurity("is_granted(['read'], request.get('_legacy_controller'))")
```
I dont think we need to protect this action as there is no route targeting it and it seems to be a util function, not a controller action that return a Response ?

rokaszygmantas replied:
it doesn't seem harmful, but it dispatches a hook internally in FormFactory, so I'd rather secure it :thinking: 

there are some routes connected to this action, e.g. this one: https://github.com/PrestaShop/PrestaShop/blob/e987584b5b30cd74cf3a7b185343cfc3a0d577fe/src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/categories.yml#L8

matks replied:
My bad, from github page I thought this was related `renderFieldAction` instead of `searchGridAction`. Now it makes sense 👍 .

By the way I'm wondering ... why do we have some controllers who provide their own `searchGridAction` and some controllers which use this generic action ?

rokaszygmantas replied:
> why do we have some controllers who provide their own searchGridAction and some controllers which use this generic action ?

I don't really know, but my guess is that the generic action was introduced later and some grids already had their implementations of search action :thinking:

matks replied:
Created issue https://github.com/PrestaShop/PrestaShop/issues/13448 to handle this

matks commented on src/PrestaShopBundle/Controller/Admin/TranslationsController.php:
```diff
@@ -49,6 +49,8 @@ class TranslationsController extends FrameworkBundleAdminController
     const controller_name = self::CONTROLLER_NAME;
 
     /**
+     * @AdminSecurity("is_granted(['read'], request.get('_legacy_controller'))")
```
I dont know whether the fact it was not secured is intentional however the returned content seems quite harmless. I dont know :/

rokaszygmantas replied:
are there any downsides of securing it?

matks replied:
Yes, if this ajax route is used to retrieve informations from other page.

For example if page Module calls this ajax route to display some informations, then a user who is granted only READ access to module page needs to have also translations page READ access to display the Module page correctly. Before, as it was unsecured ... it worked 😅

rokaszygmantas replied:
true, this is tricky :thinking:  but on the other hand, if all other pages are secured the same way, why should we skip security on this one?

matks replied:
It's rather than this is a page that was migrated in the early days of PS 1.7 when the migration plan was not clear nor clean, so it might have been unsecured on purpose and I'm afraid of breaking something that worked (and was unsecured on purpose by design).

rokaszygmantas replied:
so should I remove the security from this action?

matks replied:
I think it's better for now indeed

rokaszygmantas replied:
removed it

rokaszygmantas committed:
Remove security from translations overview action

matks commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeController.php:
```diff
@@ -367,8 +367,6 @@ public function resetLayoutsAction($themeName)
     /**
      * Show Front Office theme's pages layout customization.
      *
-     * @DemoRestricted(redirectRoute="admin_themes_index")
```
Why this removal ?

rokaszygmantas replied:
the demo permissions are checked in the action's code, it is supposed to allow showing this page with demo mode, but not allow changing values

matks commented:
@marionf @sarahdib I'm sorry, this PR requires heavy QA 😞

marionf commented:
@rokaszygmantas 

1) I have a profile with only "view" permission for Customers.
With this profile when I try to add or edit a customer, I am redirected to Catalog > Products without any information.

In legacy I had these error messages: 

![capture d'écran_1433](https://user-images.githubusercontent.com/13449658/56359666-80a04100-61e2-11e9-9ac6-1d795de9ab3f.png)

![capture d'écran_1434](https://user-images.githubusercontent.com/13449658/56359701-931a7a80-61e2-11e9-8c12-6fbeadffe6c0.png)

2) On stock overview page, I can edit quantities even with only "view" permission

3) I can edit translations even if I have only "view" permission

4) Question: I should be able to search and export with which permission ? If I have only view or only add or only edit ?

rokaszygmantas committed:
Protect stock edit actions

rokaszygmantas commented:
@marionf 

1. The problem is that in case of insufficient permissions you are redirected to a legacy page (customers list) and it doesn't display the error messages from migrated pages there. If you would change your employee default page to any migrated page - you would see the error message. Not sure how should this be fixed, ping @matks :thinking: 

2. Fixed.

3. After a discussion with @matks I've removed the security from this page as it was not secured in previous versions as well.

4. `view` permission should be enough for both search and export.

marionf commented:
@rokaszygmantas 

I have only "View" right for stock page.
If I try to change quantity for one product I have the following error:

![capture d'écran_1456](https://user-images.githubusercontent.com/13449658/56658283-84c4d680-669a-11e9-8218-a0292345d773.png)


If I try to change the quantity in bulk, I have a success message but quantity isn't saved.

![capture d'écran_1455](https://user-images.githubusercontent.com/13449658/56658313-9ad29700-669a-11e9-9cff-9dd4224dee7d.png)

Do you think we can have an error message instead ?

rokaszygmantas committed:
Fix security in bulk stock update

rokaszygmantas commented:
@marionf yes, fixed it now - it will show the same error message in bulk update. :slightly_smiling_face:

marionf commented:
Thank you @rokaszygmantas

matks commented:
Thank you @marionf for the heavy QA work
@rokaszygmantas Code review and QA are ✅but unfortunately PrettyCI is not happy about the codestyle 😅

Quetzacoalt91 committed:
Apply PHP CS fixer

Quetzacoalt91 reviewed (approved):
Rebased and applied PHP-cs-fixer. Now ready for merge.

Quetzacoalt91 commented:
Thank you @rokaszygmantas
